File: python/src/semantic_retrieval/retrieval/retriever.py
from abc import abstractmethod
from typing import Generic, Optional

from result import Result
from semantic_retrieval.access_control.access_passport import AccessPassport
from semantic_retrieval.common.types import Q, R
from semantic_retrieval.document.metadata.document_metadata_db import (
    DocumentMetadataDB,
)
from semantic_retrieval.utils.callbacks import CallbackManager


class BaseRetrieverQueryParams(Generic[Q]):
    def __init__(self, access_passport: AccessPassport, query: Q) -> None:
        self.access_passport = access_passport
        self.query = query


class BaseRetriever(Generic[R, Q]):
    def __init__(
        self,
        metadata_db: Optional[DocumentMetadataDB] = None,
        callback_manager: Optional[CallbackManager] = None,
    ) -> None:
        self.metadata_db = metadata_db
        self.callback_manager = callback_manager

    @abstractmethod
    async def retrieve_data(
        self, params: BaseRetrieverQueryParams[Q]
    ) -> Result[R, str]:
        pass

File: typescript/src/retrieval/retriever.ts
import { AccessPassport } from "../access-control/accessPassport";
import { DocumentMetadataDB } from "../document/metadata/documentMetadataDB";
import { CallbackManager, Traceable } from "../utils/callbacks";

export interface BaseRetrieverQueryParams<Q = unknown> {
  accessPassport?: AccessPassport;
  query: Q;
}

export type RetrieverResponse<R> = R extends BaseRetriever<infer RR>
  ? RR
  : never;

export type RetrieverParams<R> = R extends BaseRetriever<infer _RR, infer RP>
  ? RP
  : never;

export type RetrieverParamsQuery<P> = P extends BaseRetrieverQueryParams<
  infer Q
>
  ? Q
  : never;

export type RetrieverQuery<R> = R extends BaseRetriever<infer _RR, infer RP>
  ? RP extends BaseRetrieverQueryParams<infer QP>
    ? QP
    : never
  : never;

/**
 * Abstract base class for retrieving data R from from an underlying source based on query Q.
 */
export abstract class BaseRetriever<
  R = unknown,
  P extends BaseRetrieverQueryParams = BaseRetrieverQueryParams,
> implements Traceable
{
  metadataDB?: DocumentMetadataDB;
  callbackManager?: CallbackManager;

  constructor(
    metadataDB?: DocumentMetadataDB,
    callbackManager?: CallbackManager
  ) {
    this.metadataDB = metadataDB;
    this.callbackManager = callbackManager;
  }

  /**
   * Get the data relevant to the given query and which the current identity can access.
   * @param params The retriever query params to use for the query.
   * @returns A promise that resolves to the retrieved data.
   */
  abstract retrieveData(params: P): Promise<R>;
}
